index.bs

```diff
@@ -6,16 +6,111 @@ Status: w3c/UD
+Default Ref Status: current
 </pre>
 
 Introduction {#intro}
 =====================
 
-[[WEBDRIVER|WebDriver]] defines a protocol for introspection and remote control
-of user agents. This specification extends WebDriver by introducing
-bidirectional communication. In place of the strict command/response
-format of WebDriver, this permits events to stream from the user
-agent to the controlling software, better matching the evented nature
-of the browser DOM.
+[[WEBDRIVER2|WebDriver]] defines a protocol for introspection and
+remote control of user agents. This specification extends WebDriver by
+introducing bidirectional communication. In place of the strict
+command/response format of WebDriver, this permits events to stream
+from the user agent to the controlling software, better matching the
+evented nature of the browser DOM.
+
+Terminology {#terminology}
+==========================
+
+This specification reuses the terms defined in the
+[[!WEBDRIVER2]] specification.
+
+Transport {#transport}
+======================
+
+Message transport is provided using a [[RFC6455|WebSocket]]
+connection.
+
+A [[WEBDRIVER2#dfn-remote-ends|remote end]] has a
+<dfn>connection</dfn>, which is a network connection, and a
+<dfn>websocket url</dfn>, which is a string. These are both initially
+<code>null</code>.
+
+When data is recieved on [=connection=] it must be processed according
+to the requirements of the [[!RFC6455|WebSockets]]
+specification. When a [[RFC6455#section-6.2|WebSocket Message has
+been Received]], an implementation must [=handle an incoming
+message=]. When [[RFC6455#section-7.1.3|The WebSocket Closing
+Handshake is Started]] an implementation must [=end the
+session=].
+
+When required to <dfn>ensure a connection is established</dfn>, run
+the following steps:
+
+1.  If the [=websocket url=] is not <code>null</code>, return the
+    [=websocket url=]
+
+2.  Set up a network connection that listens on an
+    implementation-defined hostname <var>host</var> and port
+    <var>port</var>. Set the [=connection=] to this network connection.
+
+3.  Let <var>WebSocket URL</var> be the result of constructing a
+    [[RFC6455#section-3|WebSocket ws-URI]] with host
+    <var>host</var>, port <var>port</var>, and empty path an query.
+
+4.  Set the [=websocket url=] to <var>WebSocket URL</var>.
+
+5.  Return <var>WebSocket URL</var>.
+
+
+Note: Typically the host name in the above steps will be
+"<code>localhost</code>".
+
+When required to <dfn>handle an incoming message</dfn>, run
+the following steps:
+
+Issue: Figure this out.
+
+When required to <dfn>end the session</dfn>, run the following
+steps:
+
+1. Close the underlying network connection [=connection=].
+
+2. Set [=connection=] and [=WebSocket URL=] to <code>null</code>.
+
+Issue: This should also reset any internal state
+
+Note: This does not end any [[WEBDRIVER2#sessions|WebDriver
```

Comment:
Is this a different `session` mentioned above?

Reply:
Yeah, maybe this is confusing. The point is that the bidi connection has a different lifetime compared to the WebDriver session. This is necessarily true, because the client can send a close frame which will cause the connection to close, without affecting the HTTP connection. And we could force the overall session to close, but that adds coupling between the specs in a way that's not really necessary.

Reply:
I've removed "session" for now to avoid confusion.